most.js

```diff
@@ -1,42 +1,28 @@
 'use strict';
 
-const fs = require('fs');
-const path = require('path');
+const most          = require('most');
 
-const most = require('most');
-const R = require('ramda');
-const S = require('sanctuary');
+const exit0         = require('./common/exit0');
+const exit1         = require('./common/exit1');
+const join          = require('./common/join');
+const readFile      = require('./common/read-file-promise');
+const S             = require('./common/sanctuary');
 
 
-//    join :: String -> String -> String
-const join = R.curryN(2, path.join);
-
-//    readFile :: String -> String -> Promise Error String
-const readFile = R.curry((encoding, filename) =>
-  new Promise((resolve, reject) =>
-    fs.readFile(filename, { encoding: encoding }, (err, data) =>
-      err != null ? reject(err) : resolve(data)
-    )));
-
-//    concatFiles :: String -> Promise Error String
-const concatFiles = dir =>
-  R.pipe(R.pipe(readFile('utf8'), most.fromPromise),
-         R.map(S.lines),
-         R.map(R.map(join(dir))),
-         R.chain(R.pipe(R.map(readFile('utf8')), most.from)),
-         stream => stream.await(),
-         R.reduce(R.concat, '')
-  )(join(dir, 'index.txt'));
+//    concatFiles :: (String -> String) -> Promise Error String
+const concatFiles = path =>
+  most.fromPromise(readFile(path('index.txt')))
+  .map(S.lines)
+  .map(S.map(path))
+  .map(S.map(readFile))
+  .map(Promise.all.bind(Promise))
+  .awaitPromises()
```

Comment:
The problem is that the Promise settles before Most attaches handlers. In order to make sure handlers are attached, we use `most.fromPromise` immediately on the return value from `readFile`:

```js
//    readFileStream :: String -> Stream Error String
const readFileStream = S.compose(most.fromPromise, readFile);

//    concatFiles :: (String -> String) -> Promise Error String
const concatFiles = path =>
  most.fromPromise(readFile(path('index.txt'))) // :: Stream Error String
  .map(S.lines)                                 // :: Stream Error (Array String)
  .map(S.map(path))                             // :: Stream Error (Array String)
  .map(S.map(readFileStream))                   // :: Stream Error (Array (Stream Error String))
  .chain(most.mergeArray)                       // :: String Error String
  .reduce(S.concat, '')                         // :: Promise Error String
```